gdal/frmts/rdb/rdbdataset.cpp

```diff
@@ -0,0 +1,784 @@
+        {
+            SetBandInternal(this, osAttributeName, oAttribute,
+                            oAttribute.dataType(), nNumberOfLevels - 1,
+                            nNumberOfLevels, nBandIndex);
+        }
+        else
+        {
+            for(uint32_t i = 0; i < oAttribute.length; i++)
+            {
+                std::ostringstream oOss;
+                oOss << osAttributeName << '[' << i << ']';
+                SetBandInternal(this, oOss.str(), oAttribute,
+                                oAttribute.dataType(), nNumberOfLevels - 1,
+                                nNumberOfLevels, nBandIndex);
+            }
+        }
+    }
+}
+
+GDALDataset *RDBDataset::Open(GDALOpenInfo *poOpenInfo)
+{
+    if(!Identify(poOpenInfo))
+    {
+        return nullptr;
+    }
+
+    if(poOpenInfo->eAccess == GA_Update)
+    {
+        CPLError(CE_Failure, CPLE_NotSupported,
+                 "The RDB driver does not support update access to existing "
+                 "datasets.");
+        return nullptr;
+    }
+
+    if(poOpenInfo->fpL == nullptr)
+    {
+        return nullptr;
+    }
+    try
+    {
+        RDBDataset *poDS = new RDBDataset(poOpenInfo);
+        // std::swap(poDS->fp, poOpenInfo->fpL);
+        // Initialize any PAM information.
+        poDS->SetDescription(poOpenInfo->pszFilename);
+        poDS->TryLoadXML();
+
+        return poDS;
+    }
+    catch(const riegl::rdb::Error &oException)
+    {
+        CPLError(CE_Failure, CPLE_OpenFailed, "RDB error: %s, %s",
+                 oException.what(), oException.details());
+        return nullptr;
+    }
+
+    catch(const std::exception &oException)
+    {
+        CPLError(CE_Failure, CPLE_OpenFailed, "Error: %s", oException.what());
+        return nullptr;
+    }
+    catch(...)
+    {
+        CPLError(CE_Failure, CPLE_OpenFailed, "Unknown error in Open.");
+        return nullptr;
+    }
+
+    return nullptr;
+}
+int RDBDataset::Identify(GDALOpenInfo *poOpenInfo)
+{
+    const char *psHeader = reinterpret_cast<char *>(poOpenInfo->pabyHeader);
+    if(poOpenInfo->nHeaderBytes < 32)
+    {
+        return FALSE;
+    }
+
+    constexpr int kSizeOfRDBHeaderIdentifier = 32;
+    constexpr char szRDBHeaderIdentifier[kSizeOfRDBHeaderIdentifier] =
+        "RIEGL LMS RDB 2 POINTCLOUD FILE";
+
+    if(strncmp(psHeader, szRDBHeaderIdentifier, kSizeOfRDBHeaderIdentifier))
+    {
+        // A more comprehensive test could be done by the library.
+        // Should file -> library incompatibilities handled in Identify or
+        // in the Open function?
+        return FALSE;
+    }
+    return TRUE;
+}
+
+CPLErr RDBDataset::GetGeoTransform(double *padfTransform)
+{
+    padfTransform[0] = dfXMin;
+    padfTransform[1] = dfSizeOfPixel;
+    padfTransform[2] = 0;
+
+    padfTransform[3] = dfYMin;
+    padfTransform[4] = 0;
+    padfTransform[5] = dfSizeOfPixel;
+
+    return CE_None;
+}
+const char *RDBDataset::_GetProjectionRef()
+{
+    ReadGeoreferencing();
```

Comment:
I removed that function and did not put it into `GetSpatialRef()` since that function is `const` and `ReadGeoreferening()` is called in the constructor anyway.